basis/math/matrices/matrices.factor

```diff
@@ -1,197 +1,118 @@
-: n-m  ( n m -- m ) [ n-v ] with map ;
-: m-n  ( m n -- m ) [ v-n ] curry map ;
-: n*m ( n m -- m ) [ n*v ] with map ;
-: m*n ( m n -- m ) [ v*n ] curry map ;
-: n/m ( n m -- m ) [ n/v ] with map ;
-: m/n ( m n -- m ) [ v/n ] curry map ;
+DEFER: dim
+! can't define dim using this predicate for this reason,
+! unless we are going to write two versions of dim, one of which is generic
+PREDICATE: square-matrix < matrix
+    { [ well-formed-matrix? ] [ dim all-eq? ] } 1&& ;
 
-: m+   ( m m -- m ) [ v+ ] 2map ;
-: m-   ( m m -- m ) [ v- ] 2map ;
-: m*   ( m m -- m ) [ v* ] 2map ;
-: m/   ( m m -- m ) [ v/ ] 2map ;
+! really truly empty
+PREDICATE: null-matrix < matrix
+    flatten empty? ;
 
-: v.m ( v m -- v ) flip [ v. ] with map ;
-: m.v ( m v -- v ) [ v. ] curry map ;
-: m.  ( m m -- m ) flip [ swap m.v ] curry map ;
+! just full of zeroes
+PREDICATE: zero-matrix < matrix
+    dup null-matrix? [ drop f ] [ flatten [ zero? ] all? ] if ;
 
-: m~  ( m m epsilon -- ? ) [ v~ ] curry 2all? ;
+! square and full of zeroes
+PREDICATE: zero-square-matrix < square-matrix
+    { [ zero-matrix? ] [ square-matrix? ] } 1&& ;
 
-: mmin ( m -- n ) [ 1/0. ] dip [ [ min ] each ] each ;
-: mmax ( m -- n ) [ -1/0. ] dip [ [ max ] each ] each ;
-: mnorm ( m -- n ) dup mmax abs m/n ;
-: m-infinity-norm ( m -- n ) [ [ abs ] map-sum ] map supremum ;
-: m-1norm ( m -- n ) flip m-infinity-norm ;
-: frobenius-norm ( m -- n ) [ [ sq ] map-sum ] map-sum sqrt ;
+<PRIVATE
+: (nth-from-end) ( n seq -- n )
+    length 1 - swap - ; inline
 
-: cross ( vec1 vec2 -- vec3 )
-    [ [ { 1 2 0 } vshuffle ] [ { 2 0 1 } vshuffle ] bi* v* ]
-    [ [ { 2 0 1 } vshuffle ] [ { 1 2 0 } vshuffle ] bi* v* ] 2bi v- ; inline
+: nth-end ( n seq -- elt )
+    [ (nth-from-end) ] keep nth ; inline
 
-:: normal ( vec1 vec2 vec3 -- vec4 )
-    vec2 vec1 v- vec3 vec1 v- cross normalize ; inline
+: set-nth-end ( elt n seq -- )
+    [ (nth-from-end) ] keep set-nth ; inline
 
-: proj ( v u -- w )
-    [ [ v. ] [ norm-sq ] bi / ] keep n*v ;
+PRIVATE>
 
-: perp ( v u -- w )
-    dupd proj v- ;
+! Benign matrix constructors
+: <matrix> ( m n element -- matrix )
+    '[ _ _ <array> ] replicate ; inline
 
-: angle-between ( v u -- a )
-    [ normalize ] bi@ h. acos ;
+: <matrix-by> ( m n quot: ( ... -- elt ) -- matrix )
+    '[ _ _ replicate ] replicate ; inline
 
-: (gram-schmidt) ( v seq -- newseq )
-    [ dupd proj v- ] each ;
+: <matrix-by-indices> ( ... m n quot: ( ... m' n' -- ... elt ) -- ... matrix )
+    [ [ <iota> ] bi@ ] dip cartesian-map ; inline
 
-: gram-schmidt ( seq -- orthogonal )
-    V{ } clone [ over (gram-schmidt) suffix! ] reduce ;
+: <zero-matrix> ( m n -- matrix )
+    0 <matrix> ; inline
 
-: norm-gram-schmidt ( seq -- orthonormal )
-    gram-schmidt [ normalize ] map ;
+: <zero-square-matrix> ( n -- matrix )
+    dup <zero-matrix> ; inline
 
-ERROR: negative-power-matrix m n ;
+! main-diagonal matrix
+! running time is improved by 10% over the old implementation
```

Comment:
This is a good comment to have in the discussion but will quickly get dated in the source code.